### DISCUSSION TITLE
Multiple Cores on Dagster Code Location gRPC Server

### DISCUSSION THREAD

**Discussion by akan72** [dagster-io/dagster#17646] [Q&A] [answered]
Can the gRPC server that Dagster uses (`dagster api grpc`) for code locations take advantages of multiple CPU cores? We're using Dagster k8s. 

We're configured the pod running the code location to request 1000m and 1500m CPU (config below), so it should be able to burst to 1.5 cores. However it's limiting to only one core (1000m).

<img width="400" alt="image" src="https://github.com/dagster-io/dagster/assets/29241719/661208ad-f41a-48e8-844f-659c3ee95455">

**Comment by gibsondan** [comment, marked as answer]
I suspect this is because the gRPC server is using a ThreadPoolExecutor, which doesn't support multiple CPU cores due to Python / the GIL: https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_grpc/server.py#L980-L990

**Reply by akan72** [reply]
Thanks for the quick reply!

Do you have any recommendations for improving the robustness of the code location? My immediate thought is just to increase `--max-workers`. 

I see that the [default](https://github.com/dagster-io/dagster/blob/d2170b89b69d02bc57f1d5b00277e23e33e3fc17/python_modules/dagster/dagster/_cli/api.py#L472-L479) here is `None`. Does this resolve to 1 under the hood?

**Reply by gibsondan** [reply]
It uses the default Python value by default, which is "the number of processors on the machine, multiplied by 5": https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor
